tests/utils/test_qtutils.py

```diff
@@ -91,3 +125,836 @@ def test_qt_both(self, parser):
+            pyqiodev_failing.write(b'x')
+        assert str(excinfo.value) == 'Writing failed'
+
+    @pytest.mark.skipif(os.name != 'posix', reason="Needs a POSIX OS.")
+    @pytest.mark.skipif(not os.path.exists('/dev/full'),
+                        reason="Needs /dev/full.")
+    def test_write_error_real(self):
+        """Test a real write error with /dev/full on supported systems."""
+        qf = QFile('/dev/full')
+        qf.open(QIODevice.WriteOnly | QIODevice.Unbuffered)
+        dev = qtutils.PyQIODevice(qf)
+        with pytest.raises(OSError) as excinfo:
+            dev.write(b'foo')
+        qf.close()
+        assert str(excinfo.value) == 'No space left on device'
+
+    @pytest.mark.parametrize('size, chunks', [
+        (-1, [b'1234567890']),
+        (0, [b'']),
+        (3, [b'123', b'456', b'789', b'0']),
+        (20, [b'1234567890'])
+    ])
+    def test_read(self, size, chunks, pyqiodev):
+        """Test reading with different sizes.
+
+        Args:
+            size: The size to pass to read()
+            chunks: A list of expected data chunks.
+        """
+        with pyqiodev.open(QIODevice.WriteOnly) as f:
+            f.write(b'1234567890')
+        pyqiodev.open(QIODevice.ReadOnly)
+        for i, chunk in enumerate(chunks):
+            print("Expecting chunk {}: {!r}".format(i, chunk))
+            assert pyqiodev.read(size) == chunk
+
+    @pytest.mark.parametrize('method, args', [
+        ('read', []),
+        ('read', [5]),
+        ('readline', []),
+        ('readline', [5]),
+    ])
+    def test_failing_reads(self, method, args, pyqiodev_failing):
+        """Test reading with a FailingQIODevice.
+
+        Args:
+            method: The name of the method to call.
+            args: A list of arguments to pass.
+        """
+        func = getattr(pyqiodev_failing, method)
+        with pytest.raises(OSError) as excinfo:
+            func(*args)
+        assert str(excinfo.value) == 'Reading failed'
+
+
+class TestEventLoop:
+
+    """Tests for EventLoop.
+
+    Attributes:
+        loop: The EventLoop we're testing.
+    """
+
+    # pylint: disable=protected-access
+
+    def _assert_executing(self):
+        """Slot which gets called from timers to be sure the loop runs."""
+        assert self.loop._executing
+
+    def _double_exec(self):
+        """Slot which gets called from timers to assert double-exec fails."""
+        with pytest.raises(AssertionError):
+            self.loop.exec_()
+
+    def test_normal_exec(self):
+        """Test exec_ without double-executing."""
+        self.loop = qtutils.EventLoop()
+        QTimer.singleShot(100, self._assert_executing)
+        QTimer.singleShot(200, self.loop.quit)
+        self.loop.exec_()
+        assert not self.loop._executing
+
+    def test_double_exec(self):
+        """Test double-executing."""
+        self.loop = qtutils.EventLoop()
+        QTimer.singleShot(100, self._assert_executing)
```

Comment:
Hehehe, good point. I'll leave them as-is for now, and if I notice something segfaulting in this test at least I know what to look for.